text/0000-procedural-vtables.md

```diff
@@ -0,0 +1,490 @@
+(https://github.com/rust-lang/rfcs/pull/2955) we can give users the capability
+to invent their own wide pointers (and thus custom dynamically sized types).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In order to mark a trait (`MyTrait`) as using a custom vtable layout, you implement the `CustomUnsized` trait for `dyn MyTrait`.
+
+```rust
+trait MyTrait: SomeSuperTrait {
+    fn some_fn(&mut self, x: u32) -> i32;
+}
+
+impl CustomUnsized for dyn MyTrait {
+    type WidePointer = MyWidePointer;
+    // details later in this RFC
+}
+
+```
+
+`MyWidePointer` is the backing type that is going to be used for wide pointers to the given trait, so
+
+* `&dyn MyTrait`
+* `Box<dyn MyTrait>`
+* `Arc<dyn MyTrait>`
+
+and any other container types that can use wide pointers. Normally when unsizing from a concrete
+pointer like `&MyStruct` to `&dyn MyTrait` a wide pointer (that essentially is
+`(&MyStruct, &'static Vtable)`) is produced. This is currently done via compiler magic, but
+further down in this section you can see how it theoretically could be done in user code.
+Likely it will stay compiler magic out of compile-time performance reasons.
+
+All `impl`s of `MyTrait` will now use `MyWidePointer` for generating the wide pointer.
+
+You are actually generating the wide pointer, not a description of it.
+Since your `impl`'s `from` function is being interpreted in the target's environment, all target specific information will match up.
+Now, you need some information about the `impl` in order to generate your metadata (and your vtable).
+You get this information directly from the type (the `T` parameter) by adding trait bounds on it.
+
+As an example, consider the function which is what normally generates your metadata.
+
+```rust
+#[repr(C)]
+struct Pointer {
+    ptr: *mut u8,
+    vtable: &'static VTable,
+}
+
+/// For going from a `&SomeStruct<dyn Trait>` to a field of `SomeStruct`.
+unsafe impl CustomProjection for Pointer {
+    unsafe fn project(ptr: Pointer, offset: usize) -> *mut u8 {
+        ptr.ptr.offset(offset),
+    }
+}
+
+unsafe impl CustomProjectionToUnsized for Pointer {
+    /// For going from `&SomeStruct<dyn Trait>` to the unsized field
+    unsafe fn project_unsized(ptr: Pointer, offset: usize) -> Pointer {
+        Pointer {
+            ptr: ptr.ptr.offset(offset),
+            vtable: ptr.vtable,
+        }
+    }
+}
+
+/// If the `owned` flag is `true`, this is an owned conversion like
+/// in `Box<T> as Box<dyn Trait>`. This distinction is important, as
+/// unsizing that creates a vtable in the same allocation as the object
+/// (like C++ does), cannot work on non-owned conversions. You can't just
+/// move away the owned object. The flag allows you to forbid such
+/// unsizings by triggering a compile-time `panic` with an explanation
+/// for the user.
+unsafe impl<T: MyTrait> const CustomUnsize<Pointer> for T {
+    fn unsize<const owned: bool>(ptr: *mut T) -> Pointer {
```

Comment:
Hmmm. This is really interesting. I suppose you could use some totally opaque type as a placeholder for the data pointer (maybe an `extern type`?) which would be swapped out at coercion time. That would also solve the issues with projection because Rust could to do whatever it wants with that particular field...

I'll have to think about it some more but I like that direction.

Reply:
We don't actually need a type, we can create a dangling symbolic pointer (this is a value in mir interpreter!) and just use `*const T` for whatever `T` we desire. Any interaction with the pointer beyond copying it around will make CTFE complain to you at compile-time. Since this is a special case, we can actually not just tell you about "dangling pointer" but special case the diagnostic to be really helpful